pallets/market/src/lib.rs

```diff
@@ -0,0 +1,1092 @@
+			/// The account that created the pool.
+			creator: T::AccountId,
+			/// The pool id and the account ID of the pool.
+			pool_id: PoolIdOf<T>,
+			/// The id of the liquidity tokens that will be minted when assets are added to this
+			/// pool.
+			lp_token: T::CurrencyId,
+			/// The asset ids associated with the pool. Note that the order of the assets may not be
+			/// the same as the order specified in the create pool extrinsic.
+			assets: AssetPairOf<T>,
+		},
+
+		/// A successful call of the `AddLiquidity` extrinsic will create this event.
+		LiquidityMinted {
+			/// The account that the liquidity was taken from.
+			who: T::AccountId,
+			/// The id of the pool that the liquidity was added to.
+			pool_id: PoolIdOf<T>,
+			/// The amounts of the assets that were added to the pool.
+			amounts_provided: BalancePairOf<T>,
+			/// The id of the LP token that was minted.
+			lp_token: T::CurrencyId,
+			/// The amount of lp tokens that were minted of that id.
+			lp_token_minted: T::Balance,
+			/// The new total supply of the associated LP token.
+			total_supply: T::Balance,
+		},
+
+		/// A successful call of the `RemoveLiquidity` extrinsic will create this event.
+		LiquidityBurned {
+			/// The account that the liquidity token was taken from.
+			who: T::AccountId,
+			/// The id of the pool that the liquidity was taken from.
+			pool_id: PoolIdOf<T>,
+			/// The amount of the asset that was received.
+			amounts: BalancePairOf<T>,
+			/// The amount of the associated LP token that was burned.
+			burned_amount: T::Balance,
+			/// The new total supply of the associated LP token.
+			total_supply: T::Balance,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(true, "template",);
+		}
+	}
+
+	/// Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates a liquidity pool and an associated new `lp_token` asset
+		#[pallet::call_index(0)]
+		#[pallet::weight(
+			T::WeightInfo::create_pool_xyk().max(
+			T::WeightInfo::create_pool_sswap()
+		))]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			kind: PoolKind,
+			first_asset_id: T::CurrencyId,
+			first_asset_amount: T::Balance,
+			second_asset_id: T::CurrencyId,
+			second_asset_amount: T::Balance,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			Self::check_assets_allowed((first_asset_id, second_asset_id))?;
+
+			ensure!(
+				!T::DisallowedPools::contains(&(first_asset_id, second_asset_id)),
+				Error::<T>::DisallowedPool,
+			);
+
+			let lp_token = match kind {
+				PoolKind::Xyk => {
+					let lp_token = T::Xyk::create_pool(
+						sender.clone(),
+						first_asset_id,
+						first_asset_amount,
+						second_asset_id,
+						second_asset_amount,
+					)?;
+					lp_token
+				},
+				PoolKind::StableSwap => {
+					let first_decimal = Self::get_decimals(&first_asset_id)?;
+					let second_decimal = Self::get_decimals(&second_asset_id)?;
```

Comment:
Why do we care about decimals? do not fully understand this,

Reply:
you can have two tokens
USDT with 6 decimals
DAI with 18 decimals
1_000_000 == 1_000_000_000_000_000_000 should be equal, so we need to take that into account somehow

Reply:
but,.. should we care about the rate? 
I mean, I guess that I should be able to create the following, 
TokenA: Dollar
TokenB: 50cCoin representation

I guess that I should be able to create a stable_pool with ( 1 USD, 2 cents ) , Or are we restricting the liquidity we put on both sides somehow?